### PR TITLE
Allow configuration of default container resources and improve resource handling when merging components

### PR DESCRIPTION
**Pull request by amisevsk**
### What issues does this PR fix or reference?
Closes https://github.com/devfile/devworkspace-operator/issues/1133

### Is it tested? How?
1. Verify that default container resources can be configured via DWOC and that values set here are propagated to the workspace pod's containers (when a devworkspace does not specify limits/requests)
2. Verify that when merging container contributions, the total resource requirements are the sum of the default value and whatever values are defined in the contribution (i.e. container _does not_ have default values) -- see https://github.com/devfile/devworkspace-operator/pull/1131
   * For example, a workspace created from [this sample](https://raw.githubusercontent.com/devfile/devworkspace-operator/2094e37953d61664d3b23014ace2aa55f780acff/samples/code-latest.yaml) (without the PR above) should have resources equal to
      ```yaml
      memoryLimit: 1024Mi
      memoryRequest: 256Mi
      cpuLimit: 500m
      cpuRequest: 30m
      ```
        plus the default values.

Basic testing can be done using the following

* DevWorkspaceOperatorConfig:
  ```yaml
  apiVersion: controller.devfile.io/v1alpha1
  kind: DevWorkspaceOperatorConfig
  metadata:
    name: devworkspace-operator-config
  config:
    workspace:
      defaultContainerResources:
        limits:
          memory: 200Mi
          cpu: 200m
        requests:
          memory: 100Mi
          cpu: 100m

  ```

* DevWorkspace:
  ```yaml
  kind: DevWorkspace
  apiVersion: workspace.devfile.io/v1alpha2
  metadata:
    name: plain
  spec:
    started: true
    routingClass: 'basic'
    template:
      attributes:
        controller.devfile.io/storage-type: ephemeral
      components:
        - name: merge-target
          attributes:
            controller.devfile.io/merge-contribution: "true"
          container:
            image: quay.io/wto/web-terminal-tooling:next
            command:
            - "tail"
            - "-f"
            - "/dev/null"
        - name: merge-contribution
          attributes:
            controller.devfile.io/container-contribution: "true"
          container:
            image: quay.io/wto/web-terminal-tooling:next
            memoryLimit: 555Mi
            memoryRequest: 55Mi
            cpuLimit: 555m
            cpuRequest: 55m
            command:
            - "tail"
            - "-f"
            - "/dev/null"
        - name: all-defaults-needed
          container:
            image: quay.io/wto/web-terminal-tooling:next
            command:
            - "tail"
            - "-f"
            - "/dev/null"
        - name: no-defaults-needed
          container:
            image: quay.io/wto/web-terminal-tooling:next
            memoryLimit: 555Mi
            memoryRequest: 55Mi
            cpuLimit: 555m
            cpuRequest: 55m
            command:
            - "tail"
            - "-f"
            - "/dev/null"
  ```

The resulting pod should have 
* Specified resources for the `no-defaults-needed` container
* All default values for the `all-defaults-needed` container
* For the `merge-target` container, the resource requirements should be the sum of the default values (in the DWOC) and the values added by the `merge-contribution` component.

### PR Checklist

- [ ] E2E tests pass (when PR is ready, comment `/test v8-devworkspace-operator-e2e, v8-che-happy-path` to trigger)
  - [ ] `v8-devworkspace-operator-e2e`: DevWorkspace e2e test
  - [ ] `v8-che-happy-path`: Happy path for verification integration with Che

### CODE REVIEW

File: pkg/library/resources/helper.go [hunk 1]
```diff
@@ -151,24 +149,29 @@ func ProcessResources(resources *corev1.ResourceRequirements) (*corev1.ResourceR
...
 	}
 	if len(result.Requests) == 0 {
 		result.Requests = nil
 	}
 
-	return result, nil
+	return result
+}
+
+// ValidateResources validates that a corev1.ResourceRequirements is valid, i.e. that (if specified), limits are greater than requests
```
Comment (AObuchow) [inline pkg/library/resources/helper.go:223]: Resource requirement requests are still valid if they are equal to the limit.
The implementation of this function follows this rule, however the documentation and error message should probably be modified.

i.e.

`// ValidateResources validates that a corev1.ResourceRequirements is valid, i.e. that (if specified), limits are greater than or equal to requests`

File: pkg/library/resources/helper.go [hunk 2]
```diff
@@ -151,24 +149,29 @@ func ProcessResources(resources *corev1.ResourceRequirements) (*corev1.ResourceR
...
+func ValidateResources(resources *corev1.ResourceRequirements) error {
+	memLimit, hasMemLimit := resources.Limits[corev1.ResourceMemory]
+	memRequest, hasMemRequest := resources.Requests[corev1.ResourceMemory]
+	if hasMemLimit && hasMemRequest && memRequest.Cmp(memLimit) > 0 {
+		return fmt.Errorf("memory request (%s) must be less than limit (%s)", memRequest.String(), memLimit.String())
```
Comment (AObuchow) [inline pkg/library/resources/helper.go:228]: Similarly, here: `memory request (%s) must be less than or equal to limit (%s)`

File: pkg/library/resources/helper.go [hunk 3]
```diff
@@ -151,24 +149,29 @@ func ProcessResources(resources *corev1.ResourceRequirements) (*corev1.ResourceR
...
+	}
+
+	cpuLimit, hasCPULimit := resources.Limits[corev1.ResourceCPU]
+	cpuRequest, hasCPURequest := resources.Requests[corev1.ResourceCPU]
+	if hasCPULimit && hasCPURequest && cpuRequest.Cmp(cpuLimit) > 0 {
+		return fmt.Errorf("CPU request (%s) must be less than limit (%s)", cpuRequest.String(), cpuLimit.String())
```
Comment (AObuchow) [inline pkg/library/resources/helper.go:234]: `CPU request (%s) must be less than or equal to limit (%s)`*

**Review by AObuchow — CHANGES_REQUESTED** [review]
Minor comments on wording, but looks great to me. Very satisfying refactor as well :sunglasses:. Tested on OpenShift 4.12.

**Review by AObuchow — APPROVED** [review]
Approving this PR as it works as intended, though note my comments on wording.

File: pkg/library/resources/helper_test.go [hunk 4]
```diff
@@ -0,0 +1,371 @@
...
+			expected:  getResourceRequirements("250Mi", "450Mi", "250m", "450m"),
+		},
+		{
+			name:      "Does not add resources if not defined in base",
+			resources: getResourceRequirements("", "", "", ""),
+			toAdd:     getResourceRequirements("150Mi", "250Mi", "150m", "250m"),
+			expected:  getResourceRequirements("", "", "", ""),
+		},
+		{
+			// Not sure if this is ultimately the desired behavior here (maybe depends on usage)
```
[the remark this answers is not in the document]
Comment (amisevsk) [reply pkg/library/resources/helper_test.go:105]: This makes sense.

My concern here is specifically about our overloading of the meaning of "0" in resources -- sometimes it means "do not set a limit/request", and sometimes it's interpreted as "zero". It makes using `AddResources` a little more confusing, as doing
```
FilterResources(resources)
AddResources(resources, toAdd)
```
will have a different result than
```
AddResources(resources, toAdd)
FilterResources(resources)
```

File: pkg/library/resources/helper.go [hunk 5]
```diff
@@ -78,34 +79,30 @@ func ParseResourcesFromComponent(component *dw.Component) (*corev1.ResourceRequi
...
-// Adds the resource limits and requests that are set in the component "toAdd" to "resources".
+// Adds the resource limits and requests that are set in the component toAdd to resources.
 // Returns an error if the resources defined in toAdd could not be parsed.
 //
-// Note: only resources that are set in the function argument "resources" will be summed
-// with the resources set in "toAdd".
-// For example, if "resources" does not set a CPU limit but "toAdd" does set a CPU limit,
+// Note: only resources that are set in the argument resources will be summed with the resources set in toAdd.
+// For example, if resources does not set a CPU limit but toAdd does set a CPU limit,
 // the CPU limit of "resources" will remain unset.
```
[the remark this answers is not in the document]
Comment (amisevsk) [reply pkg/library/resources/helper.go:87]: Good catch, though the idea of unquoting resources here makes me realize why you quoted them in the first place :D 

I might just switch back to quoted everywhere -- it's an annoying thing about godoc:

>A func’s doc comment should explain what the function returns or, for functions called for side effects, what it does. **Named arguments or results can be referred to directly in the comment**, without any special syntax like backquotes. (A consequence of this convention is that names like a, which might be mistaken for ordinary words, are typically avoided.)

File: pkg/library/resources/helper.go (hunk 1 above)
Comment (amisevsk) [reply pkg/library/resources/helper.go:223]: Good catch, will update :+1:

File: pkg/library/resources/helper.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (amisevsk) [reply pkg/library/resources/helper.go:87]: Yup, quote is easier to read since we use "resources" as both a variable name and generic term :/

**Pull request merged**